src/datadog/tracer_telemetry.cpp

```diff
@@ -0,0 +1,211 @@
+      {"runtime_id", runtime_id_.string()},
+      {"debug", debug_},
+      {"application", nlohmann::json::object({
+                          {"service_name", span_defaults_->service},
+                          {"env", span_defaults_->environment},
+                          {"tracer_version", tracer_version_string},
+                          {"language_name", "cpp"},
+                          {"language_version", std::to_string(__cplusplus)},
+                      })},
+      // TODO: host information (os, os_version, kernel, etc)
+      {"host", nlohmann::json::object({
+                   {"hostname", hostname_},
+               })},
+  });
+}
+
+std::string TracerTelemetry::app_started(nlohmann::json&& tracer_config) {
+  auto telemetry_body = generate_telemetry_body("app-started");
+  // TODO: environment variables or finalized config details
+  telemetry_body["payload"] = nlohmann::json::object({
+      {"configuration", nlohmann::json::array({})},
+
+  });
+  // TODO: Until we figure out "configuration", above, include a
+  // JSON dump of the tracer configuration as "additional_payload".
+  telemetry_body["additional_payload"] =
+      nlohmann::json::array({nlohmann::json::object({
+          {"name", "tracer_config_json"},
+          {"value", tracer_config.dump()},
+      })});
+  auto app_started_payload = telemetry_body.dump();
+  return app_started_payload;
+}
+
+void TracerTelemetry::capture_metrics() {
+  std::time_t timepoint = std::chrono::duration_cast<std::chrono::seconds>(
+                              clock_().wall.time_since_epoch())
+                              .count();
+  for (auto& m : metrics_snapshots_) {
+    auto value = m.first.get().capture_and_reset_value();
+    if (value == 0) {
+      continue;
+    }
+    m.second.emplace_back(timepoint, value);
+  }
+}
+
+std::string TracerTelemetry::heartbeat_and_telemetry() {
+  auto batch_payloads = nlohmann::json::array();
+
+  auto heartbeat = nlohmann::json::object({
+      {"request_type", "app-heartbeat"},
+  });
+  batch_payloads.emplace_back(std::move(heartbeat));
+
+  auto metrics = nlohmann::json::array();
+  for (auto& m : metrics_snapshots_) {
+    auto& metric = m.first.get();
+    auto& points = m.second;
+    if (!points.empty()) {
+      auto type = metric.type();
+      if (type == "count") {
+        metrics.emplace_back(nlohmann::json::object({
+            {"metric", metric.name()},
+            {"tags", metric.tags()},
+            {"type", metric.type()},
+            {"points", points},
+            {"common", metric.common()},
+        }));
+      } else if (type == "gauge") {
+        // gauge metrics have a interval
+        metrics.emplace_back(nlohmann::json::object({
+            {"metric", metric.name()},
+            {"tags", metric.tags()},
+            {"type", metric.type()},
+            {"interval", 10},
+            {"points", points},
+            {"common", metric.common()},
+        }));
+      }
```

Comment:
That could be done while capturing metrics so we avoid iterating twice every 60s and it will make `heartbeat_and_telemetry` easier.

Reply:
I think we don't even use gauge currently. I assume Caleb left it in anticipating it would be used soon. Technically dead code, though, what do you think?